x/streamer/keeper/pump_stream.go

```diff
@@ -0,0 +1,370 @@
+		return math.ZeroInt(), fmt.Errorf("budget per pump is fractional: too small budget (%s) or too many pumps (%d)", budget, pumpNum)
+	}
+	randBig := GenerateUnifiedRandom(ctx, modulo.BigIntMut())
+	rand := math.NewIntFromBigIntMut(randBig)
+	return math.MinInt(rand, left), nil
+}
+
+// GenerateUnifiedRandom a unified random variable by modulo.
+func GenerateUnifiedRandom(ctx sdk.Context, modulo *big.Int) *big.Int {
+	h := sha256.New()
+	h.Write(ctx.HeaderHash())
+	// h.Write(blockTimeBytes)
+	seed := h.Sum(nil)
+
+	randomBig := new(big.Int).SetBytes(seed)
+	return new(big.Int).Mod(randomBig, modulo)
+}
+
+// ExecutePump performs the pump operation by buying tokens for a specific rollapp.
+// CONTRACT: pumpAmt is always in base denom.
+func (k Keeper) ExecutePump(
+	ctx sdk.Context,
+	pumpAmt sdk.Coin,
+	rollappID string,
+) (tokenOut sdk.Coin, err error) {
+	rollapp, found := k.rollappKeeper.GetRollapp(ctx, rollappID)
+	if !found {
+		return sdk.Coin{}, fmt.Errorf("rollapp not found: %s", rollappID)
+	}
+
+	plan, found := k.iroKeeper.GetPlanByRollapp(ctx, rollapp.RollappId)
+	if !found {
+		return sdk.Coin{}, fmt.Errorf("IRO plan not found for rollapp: %s", rollappID)
+	}
+
+	buyer := k.ak.GetModuleAddress(types.ModuleName)
+
+	var tokenOutDenom string
+	if plan.IsSettled() {
+		tokenOutDenom = plan.SettledDenom
+	} else {
+		tokenOutDenom = plan.LiquidityDenom
+	}
+	var tokenOutAmt math.Int
+
+	err = osmoutils.ApplyFuncIfNoError(ctx, func(ctx sdk.Context) error {
+		// There are several cases:
+		// - targetDenom == plan.SettledDenom =>
+		//     IRO is settled =>
+		//     RA token exists =>
+		//     every RA token has an AMM route to DYM =>
+		//     do an AMM swap using feetoken
+		//
+		// - targetDenom == plan.LiquidityDenom != BaseDenom =>
+		//     IRO is not settled =>
+		//     need to buy IRO using liquidity tokens =>
+		//     every liquidity token has an AMM route to DYM =>
+		//     do an AMM swap using feetoken =>
+		//     buy IRO using tokenOut of liquidity tokens
+		//
+		// - targetDenom == plan.LiquidityDenom == BaseDenom =>
+		//     IRO is not settled =>
+		//     need to buy IRO using liquidity tokens =>
+		//     pump amount is already in base denom =>
+		//     buy IRO using pumpAmt of BaseDenom
+
+		if tokenOutDenom == pumpAmt.Denom {
+			// This can happen only for unsettled IRO with LiquidityToken == BaseDenom
+			tokenOutAmt = pumpAmt.Amount
+		} else {
+			feeToken, err := k.txFeesKeeper.GetFeeToken(ctx, tokenOutDenom)
+			if err != nil {
+				return fmt.Errorf("get fee token for denom %s: %w", tokenOutDenom, err)
+			}
+
+			reverseRoute := reverseInRoute(feeToken.Route, tokenOutDenom)
+			tokenOutAmt, err = k.poolManagerKeeper.RouteExactAmountIn(
+				ctx,
+				buyer,
+				reverseRoute,
+				pumpAmt,        // token in
+				math.ZeroInt(), // no slippage
+			)
+			if err != nil {
+				return fmt.Errorf("route exact amount in: target denom: %s, error: %w", tokenOutDenom, err)
+			}
+		}
+
+		if !plan.IsSettled() {
```

Comment:
as my previous comment - not the right check given the new feature worked by @mtsitrin